src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1499,21 +1506,29 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
                         diagnostics.Add(ErrorCode.ERR_NoMainInClass, mainType.Locations.First(), mainType);
                     }
                 }
-                else if (viableEntryPoints.Count > 1)
+                else if (nonAsyncMainEntryPointCount == 1) 
+                {
+                    entryPoint = viableEntryPoints.Where(c => !c.IsAsync).Single();
+                }
+                else if (nonAsyncMainEntryPointCount == 0 && asyncMainEntryPointCount == 1)
+                {
+                    diagnostics.AddRangeAndFree(possibleAsyncMainDiagnostics);
+                    entryPoint = viableEntryPoints.Single();
+                }
+                else if (nonAsyncMainEntryPointCount == 0 && !asyncOk)
+                {
+                    diagnostics.AddRangeAndFree(possibleAsyncMainDiagnostics);
+                }
+                else
                 {
                     viableEntryPoints.Sort(LexicalOrderSymbolComparer.Instance);
                     var info = new CSDiagnosticInfo(
                          ErrorCode.ERR_MultipleEntryPoints,
                          args: Array.Empty<object>(),
                          symbols: viableEntryPoints.OfType<Symbol>().AsImmutable(),
                          additionalLocations: viableEntryPoints.Select(m => m.Locations.First()).OfType<Location>().AsImmutable());
-
                     diagnostics.Add(new CSDiagnostic(info, viableEntryPoints.First().Locations.First()));
                 }
```

Comment:
`possibleAsyncMainDiagnostics` is not freed in the last `else`. Consider replacing all the `diagnostics.AddRangeAndFree(possibleAsyncMainDiagnostics)` with `diagnostics.AddRange(possibleAsyncMainDiagnostics)` and free `possibleAsyncMainDiagnostics` explicitly at the end of the method.